### PR TITLE
File menu save open

### PR DESCRIPTION
**Pull request by cle1000**
download working, upload only for small files, do you have any hints ?

### CODE REVIEW

File: mitmproxy/web/app.py [hunk 1]
```diff
@@ -157,6 +159,26 @@ def get(self):
...
         ))
 
+class DumpFlows(RequestHandler):
+    def get(self):
+        self.set_header("Content-Description", "File Transfer")
+        self.set_header("Cache-Control", "no-cache, no-store, must-revalidate")
+        self.set_header("Content-Disposition", "attachment; filename=flows")
+        self.set_header("Content-Type", "application/octet-stream")
+        self.set_header("Content-Transfer-Encoding", "binary")
+
```
Comment (cle1000) [inline mitmproxy/web/app.py]: i am not sure about those headers, i dont think that i need all, what do you think?

File: mitmproxy/web/app.py [hunk 2]
```diff
@@ -176,6 +176,7 @@ def get(self):
         sio.close()
 
     def post(self):
+        # self.state.clear()
```
Comment (cle1000) [inline mitmproxy/web/app.py]: what is the best way to stop mitmweb from recording, i think we also need to resume recording when a user clicks the menuitem 'new'

File: mitmproxy/web/app.py (hunk 1 above)
Comment (mhils) [inline mitmproxy/web/app.py]: Where does this list come from? :smile:
It feels like you can remove everything but `Content-Disposition` and `Content-Type`.

File: mitmproxy/web/app.py [hunk 3]
```diff
@@ -157,6 +159,27 @@ def get(self):
...
 
+class DumpFlows(RequestHandler):
+    def get(self):
+        self.set_header("Content-Description", "File Transfer")
+        self.set_header("Cache-Control", "no-cache, no-store, must-revalidate")
+        self.set_header("Content-Disposition", "attachment; filename=flows")
+        self.set_header("Content-Type", "application/octet-stream")
+        self.set_header("Content-Transfer-Encoding", "binary")
+
+        sio = StringIO()
```
Comment (mhils) [inline mitmproxy/web/app.py]: If you want to make @dufferzafar's Python 3 migration easier, this should be BytesIO (`from io import BytesIO`) :smiley:.

File: mitmproxy/web/app.py (hunk 2 above)
Comment (mhils) [inline mitmproxy/web/app.py]: Just calling `.clear()` here is the best way for now. We need to fix that later.

File: web/src/js/actions.js [hunk 4]
```diff
@@ -117,6 +117,13 @@ export var FlowActions = {
...
     clear: function(){
         $.post("/clear");
+    },
+    download: () => window.location = "/flows/dump",
+    upload: (file) => {
+        var filereader = new FileReader();
+        filereader.file = file;
+        filereader.onload = (e) => {$.post("/flows/dump", e.target.result); e.preventDefault();};
+        filereader.readAsBinaryString(file);
     }
```
Comment (mhils) [inline web/src/js/actions.js]: This looks good. At some point we probably want to kick jQuery out and use the HTML5 fetch API directly.
Do you want to include this right away?

https://stackoverflow.com/questions/36067767/how-do-i-upload-a-file-with-the-html5-js-fetch-api

File: web/src/js/components/header.js [hunk 5]
```diff
@@ -344,12 +344,19 @@ var FileMenu = React.createClass({
         }
     },
     handleOpenClick: function (e) {
+        $('#uploadFileInput').trigger('click');
```
Comment (mhils) [inline web/src/js/components/header.js]: This is not how React works :wink:

File: web/src/js/components/header.js [hunk 6]
```diff
@@ -344,12 +344,19 @@ var FileMenu = React.createClass({
...
+        e.preventDefault();
+    },
+    handleOpenFile: function (e) {
+        if (e.target.files.length > 0) {
+            FlowActions.upload(e.target.files[0]);
+            $('#uploadFileInput').val("");
```
Comment (mhils) [inline web/src/js/components/header.js]: Same. No jQuery please. :-)

File: web/src/js/store/store.js [hunk 7]
```diff
@@ -118,7 +118,8 @@ _.extend(LiveStoreMixin.prototype, {
                     this.handle_fetch(message.data);
                 }.bind(this))
                 .fail(function () {
-                    EventLogActions.add_event("Could not fetch " + this.type);
+                    //EventLogActions.add_event("Could not fetch " + this.type);
+                    console.log("Could not fetch " + this.type); // store.js:121 Uncaught ReferenceError: EventLogActions is not defined
```
Comment (mhils) [inline web/src/js/store/store.js]: `import { EventLogActions } from "./actions.js"` ?

File: mitmproxy/web/app.py [hunk 8]
```diff
@@ -157,6 +159,27 @@ def get(self):
...
+        fw = FlowWriter(sio)
+        for f in self.state.flows:
+            fw.add(f)
+        self.write(sio.getvalue())
+
+        sio.close()
+
+    def post(self):
+        # self.state.clear()
```
Comment (mhils) [inline mitmproxy/web/app.py]: Just calling .clear() here is the best way for now. We need to fix that later.

**Comment by mhils** [comment]
> download working, upload only for small files, do you have any hints ?

Not really - any error message?
(maybe things improve if we use `fetch`?)

File: web/src/js/actions.js (hunk 4 above)
Comment (cle1000) [inline web/src/js/actions.js]: i changed it to the fetch html5 api, but now i dont have the _xsrf cookie out of the box, i added it to the request but now i get the following warning message form tronado:

`WARNING:tornado.general:403 POST /flows/dump?_xsrf=2%7Ca7b2e0dc%7C0e4ba265b3d5d563622e7dbeb8945579%7C1464868825 (::1): XSRF cookie does not match POST argument`

**Comment by cle1000** [comment]
thats the error message, i canno test it with the fetch api because it is not working :( 
```
HTTPServerRequest(protocol='http', host='localhost:8081', method='POST', uri='/flows/dump?_xsrf=2%7Ca7b2e0dc%7C0e4ba265b3d5d563622e7dbeb8945579%7C1464868825', version='HTTP/1.1', remote_ip='::1', headers={'Origin': 'http://localhost:8081', 'Content-Length': '433378', 'Accept-Language': 'en-US,en;q=0.8,de;q=0.6', 'Accept-Encoding': 'gzip, deflate', 'Host': 'localhost:8081', 'Accept': '*/*', 'User-Agent': 'Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/51.0.2704.63 Safari/537.36', 'Connection': 'keep-alive', 'X-Requested-With': 'XMLHttpRequest', 'Pragma': 'no-cache', 'Cache-Control': 'no-cache', 'Referer': 'http://localhost:8081/', 'Content-Type': 'application/x-www-form-urlencoded; charset=UTF-8', 'Cookie': '_pk_id.107.1fff=115af219c0272447.1452468128.1.1452468128.1452468128.; _ga=GA1.1.1877128948.1451853567; _xsrf=2|a7b2e0dc|0e4ba265b3d5d563622e7dbeb8945579|1464868825'})
Traceback (most recent call last):
  File "/home/cle1000/git_projects/master/mitmproxy/venv/lib/python2.7/site-packages/tornado/web.py", line 1443, in _execute
    result = method(*self.path_args, **self.path_kwargs)
  File "/home/cle1000/git_projects/master/mitmproxy/mitmproxy/web/app.py", line 178, in post
    self.state.load_flows(FlowReader(bio).stream())
  File "/home/cle1000/git_projects/master/mitmproxy/mitmproxy/web/__init__.py", line 89, in load_flows
    super(WebState, self).load_flows(flows)
  File "/home/cle1000/git_projects/master/mitmproxy/mitmproxy/flow.py", line 599, in load_flows
    self.flows._extend(flows)
  File "/home/cle1000/git_projects/master/mitmproxy/mitmproxy/flow.py", line 519, in _extend
    self._list.extend(flows)
  File "/home/cle1000/git_projects/master/mitmproxy/mitmproxy/flow.py", line 1235, in stream
    raise FlowReadException("Invalid data format.")
FlowReadException: Invalid data format.
ERROR:tornado.access:500 POST /flows/dump?_xsrf=2%7Ca7b2e0dc%7C0e4ba265b3d5d563622e7dbeb8945579%7C1464868825 (::1) 44.03ms
```

File: mitmproxy/web/app.py [hunk 9]
```diff
@@ -367,7 +388,7 @@ def __init__(self, master, debug, wauthenticator):
         settings = dict(
             template_path=os.path.join(os.path.dirname(__file__), "templates"),
             static_path=os.path.join(os.path.dirname(__file__), "static"),
-            xsrf_cookies=True,
+            xsrf_cookies=False,
```
Comment (mhils) [inline mitmproxy/web/app.py]: (this needs to be fixed before we merge this)

**Comment by mhils** [comment]
The file upload borks on the encoding:
```
>>> print(posted_file[4750:4780])
">Test your âfree willâ�
>>> print(actual_file[4750:4780])
">Test your “free will” at
```

**Comment by mhils** [comment]
Fixed. @cle1000, can you do a final review and then we can merge this? :smiley:

**Pull request merged**